Implement gather primitive for ProcessGroupNCCL

lishanYang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62918
* #61437
* __->__ #61436

NCCL doesn’t provide primitives for gather, and these need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.
1. In test_c10d_nccl.py, a test function ```test_gather_ops``` is added.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

Commit:
Implement gather primitive for ProcessGroupNCCL

NCCL doesn’t provide primitives for gather, and these need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.
1. In test_c10d_nccl.py, a test function ```test_gather_ops``` is added.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


NCCL doesn’t provide primitives for gather, and these need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.
1. In test_c10d_nccl.py, a test function ```test_gather_ops``` is added.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
+        for idx, ls in enumerate(output_ts):
+            if idx == self.rank:
+                self.assertEqual(output_ts[idx], golden_ts[idx])
+            else:
+                for i in range(self.num_gpus):
+                    self.assertEqual(output_ts[idx][i], 0)
```
Don't need verification for non-root ranks.

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
             allgather_base(output_t, tensor)
 
+    @requires_nccl()
+    def test_gather_ops(self):
```
I think we need additional unit tests here similar to what we have in https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py. Look at the following tests in that file: [test_gather_checks](https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py#L809), [test_gather_basics](https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py#L914) and [test_gather_stress](https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py#L957). We should have similar tests in NCCL.

pritamdamania87 replied:
This test is under `ProcessGroupNCCLTest` which had only a single process. We should also have tests under `class CommTest` which tests multiple processes using NCCL.

pritamdamania87 replied:
Need to address my last comment here to ensure we have tests that run with multiple processes.

pritamdamania87 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -807,6 +807,47 @@ void recv(
+      NCCL_CHECK(ncclRecv(recvbuff, count, type, r, comm, stream));
+    }
+  }
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
```
nit: Shouldn't we have the send before the recv? I don't think it has any correctness issues since NCCL would handle it, but it is awkward that the root rank calls recv before send.

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
+        output_ts = [[] for _ in range(self.num_gpus)]
+        for idx, ls in enumerate(output_ts):
+            for _ in range(self.world_size * self.num_gpus):
+                ls.append(torch.tensor([0]).cuda(idx))
```
nit: use `torch.empty(1)` instead.

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
+        # init input
+        tensors = []
+        for i in range(self.num_gpus):
+            tensors.append(torch.tensor([self.rank]).cuda(i))
```
We should also test a case where the input tensor is of different length on each rank.

pritamdamania87 replied:
I don't see a test where we use different lengths of tensors?

pritamdamania87 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -420,6 +420,47 @@ def allgather_base(output_t, input_t):
+            for _ in range(self.world_size * self.num_gpus):
+                ls.append(torch.tensor([0]).cuda(idx))
+
+        gather(output_ts, tensors, self.rank)
```
I think it is valuable to test the underlying ProcessGroup API, but that API is not public to PyTorch users. We should test this API more thoroughly: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2126 because it is what users use.

pritamdamania87 replied:
We should test the public API as well, as I mentioned in my comment above^

rohan-varma replied:
@pritamdamania87  @lishanYang I think there are already pretty comprehensive tests for the public API for scatter/gather here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/distributed_test.py#L2635 (similar tests exist for `gather` in `distributed_c10d`). 

Seems like all we need to do to enable that coverage of the public API is remove the skip decorators for NCCL.

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


NCCL doesn’t provide primitives for gather, and these need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.
1. In test_c10d_nccl.py, a test function ```test_gather_ops``` is added.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
Can we also add that gather now works with the NCCL backend to the docs? Thanks!

pritamdamania87 commented:
> Can we also add that gather now works with the NCCL backend to the docs? Thanks!

@lishanYang You need to update this file for the docs: https://github.com/pytorch/pytorch/blob/master/docs/source/distributed.rst

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


NCCL doesn’t provide primitives for gather, and these need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.
1. In test_c10d_nccl.py, a test function ```test_gather_ops``` is added.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

[ghstack-poisoned]

1013801464 commented:
The file `ProcessGroupNCCL.cpp` contains `flatten_for_scatter_gather`, but it goes wrong for non-root nodes, because `outputTensors` is an empty vector but `inputTensors` is not.
```
c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::gather(
    std::vector<std::vector<at::Tensor>>& outputTensors,
    std::vector<at::Tensor>& inputTensors,
    const GatherOptions& opts) {
  check_gpu_tensors(inputTensors);
  auto outputFlattened =
      flatten_for_scatter_gather(outputTensors, inputTensors, size_);
...
```

Function `flatten_for_scatter_gather`: 
```
std::vector<at::Tensor> flatten_for_scatter_gather(
    std::vector<std::vector<at::Tensor>>& tensor_lists,
    std::vector<at::Tensor>& other,
    size_t world_size) {
  if (tensor_lists.size() != other.size()) {
    TORCH_CHECK(false,
        "Tensor list operands to scatter/gather must have the same length");
  }
...
```
Sorry to bother you, I do need to use `gather`.
